### DISCUSSION TITLE
Warn about covariant mutable array usage.

### DISCUSSION THREAD

**Discussion by be5invis** [dotnet/csharplang#331] [General]
**Γ ⊢ A \<: B** should NOT imply **Γ ⊢ A[] \<: B[]**, unless the array is immutable.
Reflect: https://github.com/dotnet/roslyn/issues/4211

**Comment by jnm2** [comment]
What are you proposing, taking backwards compatibility into account?

**Comment by be5invis** [comment]
@jmm2 Maybe a warning about assigning a parent class to a subclass array?

**Comment by jnm2** [comment]
@be5invis Roslyn analyzer or C# compiler warning? And what about contravariance? One errors on read, the other errors on write.

**Comment by be5invis** [comment]
@jnm2 Both are okay.

**Comment by svick** [comment]
@be5invis 

Are you talking about code like this?

```c#
class Base {}
class Derived : Base {}

…

Base[] b = new Base[1];
Derived[] d = b;
```

Because that currently reports an error ("CS0266 Cannot implicitly convert type 'Base[]' to 'Derived[]'. An explicit conversion exists (are you missing a cast?)"). It compiles if you add a cast (i.e. `Derived[] d = (Derived[])b;`), but presumably such cast means you know what you're doing, so a warning is not necessary.

**Comment by gafter** [comment]
The language specification generally does not mandate warnings. It sounds more like a compiler or analyzer request, which would be more appropriate in the Roslyn repo.

**Comment by hez2010** [comment]
This also prevents a bunch of possibilities of JIT optimization due to covariant arrays. 
IMO arrays should be made invariant by default, no one writes

```csharp
Base[] array = new Derived[10];
array[0] = new Base();
```

**Reply by CyrusNajmabadi** [reply]
This can already be done as a Roslyn analyzer if you want to block this in your own code.

**Reply by hez2010** [reply]
But the JIT compiler can't assume arrays are invariant so related optimizations cannot be performed.

**Reply by CyrusNajmabadi** [reply]
Adding a warning won't change that.  Even with a warning, the compiler wouldn't be able to make such an assumption.

**Reply by hez2010** [reply]
I think the compiler may emit an attribute like `NoConvariantCastArray` at assembly level to help JIT understand there's no covariance on arrays in an assembly.

**Reply by ufcpp** [reply]
https://github.com/dotnet/runtime/pull/58696

**Reply by ufcpp** [reply]
https://github.com/dotnet/runtime/issues/23393
